### ISSUE TITLE
WinAppDriver Azure Pipelines Task stop command is flaky

### ISSUE THREAD

**Issue by jalkire** [microsoft/WinAppDriver#765]
We have recently integrated WinAppDriver into our builds at [Accessibility Insights for Windows](https://github.com/Microsoft/accessibility-insights-windows). We've been seeing periodic errors in relation to stopping WinAppDriver in our build pipeline. The YAML we use is as follows:
```yaml
  - task: WinAppDriver.winappdriver-pipelines-task.winappdriver-pipelines-task.Windows Application Driver@0
    displayName: 'Stop - WinAppDriver'
    inputs:
      OperationType: Stop
```
This usually works as expected, but sometimes fails with:
```
  | ##[error]The process cannot access the file 'D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\winappdriverlog.txt' because it is being used by another process. |  
```
Full log output at end of issue.

Do you know if there is something we may be doing that causes this error? Or is it an issue with the build task itself?

```
  | ##[debug]Evaluating condition for step: 'Stop - WinAppDriver' |  
  | ##[debug]Evaluating: SucceededNode() |  
  | ##[debug]Evaluating SucceededNode: |  
  | ##[debug]=> True |  
  | ##[debug]Result: True |  
  | Starting: Stop - WinAppDriver |  
  | ============================================================================== |  
  | Task         : WinAppDriver |  
  | Description  : Use this extention to integrate Windows Application Driver (WinAppDriver) into your Azure Pipeline. |  
  | Version      : 0.0.1 |  
  | Author       : Microsoft |  
  | Help         : Checkout the WinAppDriver GitHub Repo to learn more - https://github.com/Microsoft/WinAppDriver |  
  | ============================================================================== |  
  | ##[debug]VstsTaskSdk 0.11.0 commit 7ff27a3e0bdd6f7b06690ae5f5b63cb84d0f23f4 |  
  | ##[debug]Entering D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\WADPowershell.ps1. |  
  | ##[debug]Loading resource strings from: D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\task.json |  
  | ##[debug]Loaded 0 strings. |  
  | ##[debug]SYSTEM_CULTURE: 'en-US' |  
  | ##[debug]INPUT_OPERATIONTYPE: 'Stop' |  
  | ##[debug]INPUT_WADARGUMENTS (empty) |  
  | ##[debug]INPUT_AGENTRESOLUTION: 'Default' |  
  | Successfully closed WinAppDriver on Agent |  
  |   |  
  | Printing WinAppDriver Console logs: |  
  |   |  
  | ##[debug]Caught exception from task script. |  
  | ##[debug]Error record: |  
  | ##[debug]Out-File : The process cannot access the file 'D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\winappdriverlog.txt' because it is being used by another process. |  
  | ##[debug]At D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\WADPowershell.ps1:41 char:2 |  
  | ##[debug]+     Out-File -Encoding Unicode -append winappdriverlog.txt |  
  | ##[debug]+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ |  
  | ##[debug]    + CategoryInfo          : OpenError: (:) [Out-File], IOException |  
  | ##[debug]    + FullyQualifiedErrorId : FileOpenFailure,Microsoft.PowerShell.Commands.OutFileCommand |  
  | ##[debug] |  
  | ##[debug]Script stack trace: |  
  | ##[debug]at <ScriptBlock>, D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\WADPowershell.ps1: line 41 |  
  | ##[debug]at <ScriptBlock>, <No file>: line 1 |  
  | ##[debug]at <ScriptBlock>, <No file>: line 22 |  
  | ##[debug]at <ScriptBlock>, <No file>: line 18 |  
  | ##[debug]at <ScriptBlock>, <No file>: line 1 |  
  | ##[debug]Exception: |  
  | ##[debug]System.IO.IOException: The process cannot access the file 'D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\winappdriverlog.txt' because it is being used by another process. |  
  | ##[debug]   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) |  
  | ##[debug]   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost) |  
  | ##[debug]   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share) |  
  | ##[debug]   at System.Management.Automation.PathUtils.MasterStreamOpen(PSCmdlet cmdlet, String filePath, Encoding resolvedEncoding, Boolean defaultEncoding, Boolean Append, Boolean Force, Boolean NoClobber, FileStream& fileStream, StreamWriter& streamWriter, FileInfo& readOnlyFileInfo, Boolean isLiteralPath) |  
  | ##[error]The process cannot access the file 'D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\winappdriverlog.txt' because it is being used by another process. |  
  | ##[debug]Processed: ##vso[task.logissue type=error]The process cannot access the file 'D:\a\_tasks\Windows Application Driver_f7c56e2a-d1f1-4cff-b5d3-a32c9516c6e0\0.0.1\winappdriverlog.txt' because it is being used by another process. |  
  | ##[debug]Processed: ##vso[task.complete result=Failed] |  
  | Finishing: Stop - WinAppDriver
```

**Comment by hassanuz** [comment]
Thanks for reporting this @jalkire, I'm investigating into it.

**Comment by hassanuz** [comment]
@jalkire - this appears to be an I/O error that happens on edge cases, notably with the use of private agents.

I _just_ pushed out a new build now which hopefully mitigates this issue - though since I haven't been able to repro the issue myself, I am unable to validate. 

Could you try the following, 

1) Confirm that the WinAppDriver task version is 0.0.2. 
1) Check to see if the issue still persists on this latest version **making no changes to the YAML**.  
1) If the issue does still persist, could you try adjusting the "Start - WinAppDriver" YAML to include the following launch argument: 

```
steps:
- task: WinAppDriver.winappdriver-pipelines-task.winappdriver-pipelines-task.Windows Application Driver@0
  displayName: 'Start - WinAppDriver'
  inputs:
    WADArguments: 4723
    AgentResolution: 1080p

```
Let me know if any of these helps resolve it. If possible, please share more details about your agent's environment (type, OS, build version, etc.) as that would very helpful in repro'ing. 

Thanks!

**Comment by alfonso-mireles** [comment]
@hassanuz It seems that with this latest update (ver 0.0.1.6) I am not getting the following error when executing the WinAppDriver Stop command

`##[error]Cannot find path 'C:\winappdriverlog.txt' because it does not exist.`

This Pipeline had been running successfully until this morning, and it coincides with the update of the latest version.

**Comment by jalkire** [comment]
@jalkire Our most recent build ran with the 0.0.2 version of the task and did not hit this issue. I will let you know if we see it again; if not, consider it fixed. Thanks!
As for agent info, we were running on what I believe is a default Azure Pipelines agent using the windows-2019-vs2019 image. Its agent version was 2.154.3.

**Comment by hassanuz** [comment]
Thanks, do keep me posted @jalkire ! If all looks good, I will close this issue within 7 days.

**Comment by hassanuz** [comment]
> `##[error]Cannot find path 'C:\winappdriverlog.txt' because it does not exist.`

Are you still seeing this error? How consistently are you seeing it? Thanks

**Comment by alfonso-mireles** [comment]
I see that error every single run.

**Comment by hassanuz** [comment]
@alfonso-mireles - what type of agent are you running on (self-hosted, private, Windows version)? Are you passing in any arguments?

**Comment by alfonso-mireles** [comment]
@hassanuz Self-Hosted on a VM. Windows 10. No arguments being passed. 
I never encountered this issue until the latest change on the pipeline Task

**Comment by hassanuz** [comment]
> @hassanuz Self-Hosted on a VM. Windows 10. No arguments being passed.
> I never encountered this issue until the latest change on the pipeline Task

Investigating into it. Please reach out to me via hassan.uraizee@microsoft.com on the chance that I may need to provide a workaround Build.

**Comment by hassanuz** [comment]
Hi @alfonso-mireles, I pushed out a minor update - can you confirm if the issue is now resolved? 

Thanks

**Comment by alfonso-mireles** [comment]
@hassanuz Got the update from Friday. But the task is still finishes with the same Error

**Comment by hassanuz** [comment]
I'm closing this ticket as the original issue was resolved. @alfonso-mireles - please reach out to me via email for us to to directly look at your particular case. Thanks.

**Comment by rubberduck203** [comment]
Closing this issue is a little premature IMO. The fix for this issue introduced a new bug that we're also seeing.
Also self hosted.

**Comment by horoxix** [comment]
> Closing this issue is a little premature IMO. The fix for this issue introduced a new bug that we're also seeing.
> Also on a self hosted agent.
> My extension settings are saying 0.0.1.7 is the latest version.

I am also blocked by this new issue.
Thank you

**Comment by Shakevg** [comment]
Faced with the same issue on self-hosted Agent Windows 10. Did anybody find a solution?

**Comment by Shakevg** [comment]
One workaround that I found: Do not fail step if error appeared on stop WinAppDriver, that helps to continue execution
![image](https://user-images.githubusercontent.com/42133199/148053295-7fa8c99f-daf5-42bf-8cf0-1be60c175953.png)
